pkg/executor/openstack.go

```diff
@@ -0,0 +1,561 @@
+		Name:      instanceName,
+		FlavorRef: flavorID,
+		ImageRef:  image,
+	}
+
+	serverOptsExt := keypairs.CreateOptsExt{
+		CreateOptsBuilder: serverOpts,
+		KeyName:           kricoKeypairName,
+	}
+
+	instance, err := servers.Create(os.client, serverOptsExt).Extract()
+
+	if err != nil {
+		return nil, errors.Wrap(err, "unable to create instance")
+	}
+
+	log.Debugf("Scheduled instance %s creation", instance.ID)
+
+	err = servers.WaitForStatus(os.client, instance.ID, "ACTIVE", 60)
+
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't launch instance")
+	}
+
+	log.Debugf("Launched instance %s", instance.ID)
+
+	associateOpts := floatingips.AssociateOpts{
+		FloatingIP: floatingIP,
+	}
+
+	err = floatingips.AssociateInstance(os.client, instance.ID, associateOpts).ExtractErr()
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't associate floating ip to instance")
+	}
+
+	log.Debugf("Associated %q floating ip", floatingIP)
+
+	remoteConfig := RemoteConfig{
+		User:    os.config.User,
+		KeyPath: os.config.SSHKeyPath,
+		Port:    22,
+	}
+
+	log.Debug("Waiting for instance to boot up")
+
+	// Wait while to ensure that everything booted up!
+	time.Sleep(time.Second*bootUpTimeOut)
+
+	type serverAttributesExt struct {
+		servers.Server
+		extendedserverattributes.ServerAttributesExt
+	}
+
+	var extendedAttributes serverAttributesExt
+
+	err = servers.Get(os.client, instance.ID).ExtractInto(&extendedAttributes)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't get instance extended attributes")
+	}
+
+	os.config.HypervisorInstanceName = extendedAttributes.InstanceName
+
+	remote, err := NewRemote(floatingIP, remoteConfig)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't create remote executor")
+	}
+	log.Debug("Created remote executor")
+
+	_, err = remote.Execute(command)
+	if err != nil {
+		return nil, errors.Wrap(err, "couldn't start remote task")
+	}
+
+	outputDirectory, err := createOutputDirectory(command, "openstack")
+	if err != nil {
+		return nil, errors.Wrapf(err, "cannot create output directory for command: %q", command)
+	}
+
+	stdoutFile, stderrFile, err := createExecutorOutputFiles(outputDirectory)
+	if err != nil {
+		removeDirectory(outputDirectory)
+		return nil, errors.Wrapf(err, "cannot create output files for command: %q", command)
+	}
+
+	stdoutFileName := stdoutFile.Name()
+	stderrFileName := stderrFile.Name()
+	stdoutFile.Close()
+	stderrFile.Close()
+
+	taskHandle := &OpenstackTaskHandle{
+		command:        command,
+		hostIP:         floatingIP,
+		stdoutFilePath: stdoutFileName,
+		stderrFilePath: stderrFileName,
+		instance:       instance.ID,
+		os:             &os,
+		running:        true,
+		requestStop:    make(chan struct{}),
+		requestDelete:  make(chan struct{}),
+		stopped:        make(chan struct{}),
+		deleted:        make(chan struct{}),
+	}
+
+	taskWatcher := &openstackWatcher{
+		instance:      instance.ID,
+		client:        os.client,
+		running:       &taskHandle.running,
```

Comment:
I haven't noticed it before, but: https://blog.golang.org/share-memory-by-communicating. Do you think it can be done?

Reply:
We don't have problem with sharing memory in this code. Threads do not need to wait for each other.